rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/wait.hpp

```diff
@@ -36,344 +36,238 @@ wait(
+        subscriptions->subscribers[i] = nullptr;
       }
-      DDS::ReturnCode_t retcode = dds_wait_set->detach_condition(read_condition);
-      if (retcode != DDS::RETCODE_OK) {
-        RMW_SET_ERROR_MSG("Failed to get detach condition from wait set");
+
+      if (dds_wait_set->detach_condition(info->read_condition_) != DDS::RETCODE_OK) {
+        RMW_SET_ERROR_MSG("failed to detach condition");
         return RMW_RET_ERROR;
       }
     }
   }
 
-  // set guard condition handles to zero for all not triggered conditions
+  // reset guard condition for all untriggered conditions
   if (guard_conditions) {
     for (size_t i = 0; i < guard_conditions->guard_condition_count; ++i) {
-      DDS::Condition * condition =
-        static_cast<DDS::Condition *>(guard_conditions->guard_conditions[i]);
+      auto condition = static_cast<DDS::Condition *>(guard_conditions->guard_conditions[i]);
       if (!condition) {
-        RMW_SET_ERROR_MSG("condition handle is null");
+        RMW_SET_ERROR_MSG("guard_condition is null");
         return RMW_RET_ERROR;
       }
 
-      // search for guard condition in active set
+      // reset the guard condition
       ::CORBA::ULong j = 0;
-      for (; j < active_conditions->length(); ++j) {
-        if ((*active_conditions)[j] == condition) {
-          DDS::GuardCondition_var guard = DDS::GuardCondition::_narrow(condition);
-          DDS::ReturnCode_t status = guard->set_trigger_value(false);
-          if (status != DDS::RETCODE_OK) {
-            RMW_SET_ERROR_MSG("failed to set trigger value");
-            return RMW_RET_ERROR;
-          }
-          break;
+      while (j < active_conditions->length() && (*active_conditions)[j] != condition) { ++j; }
+      if (j < active_conditions->length()) {
+        DDS::GuardCondition_var guard = DDS::GuardCondition::_narrow(condition);
+        if (guard->set_trigger_value(false) != DDS::RETCODE_OK) {
+          RMW_SET_ERROR_MSG("failed to set trigger value");
+          return RMW_RET_ERROR;
         }
+      } else {
+        guard_conditions->guard_conditions[i] = nullptr;
       }
-      // if guard condition is not found in the active set
-      // reset the guard handle
-      if (!(j < active_conditions->length())) {
-        guard_conditions->guard_conditions[i] = 0;
-      }
-      DDS::ReturnCode_t retcode = dds_wait_set->detach_condition(condition);
-      if (retcode != DDS::RETCODE_OK) {
-        RMW_SET_ERROR_MSG("Failed to get detach condition from wait set");
+
+      if (dds_wait_set->detach_condition(condition) != DDS::RETCODE_OK) {
+        RMW_SET_ERROR_MSG("failed to detach condition");
         return RMW_RET_ERROR;
       }
     }
   }
-/*  TODO: uncommnet this block when type support is ready and service is properly implemented.
-  // set service handles to zero for all not triggered conditions
+
+  // reset service for all untriggered conditions
+/* TODO: uncommnet this block when type support is ready.
```

Comment:
It will crash CI tests.

Reply:
I don't think we are opposed to merging if there is a test issue as long as it can be explained or handled in a temporary way as Adam had suggested. The problem right now is that we only really have one test and breaking that one puts us in the blind if we accept it.